File: mkan_customization/mkan_customization/report/timesheet_monthly_summary_per_project/timesheet_monthly_summary_per_project.py
import frappe
from frappe import _


def execute(filters=None):
	if not filters:
		filters = {}

	columns = get_columns()
	data    = get_data(filters)
	return columns, data


# ---------------------------------------------------------------------------
# Columns
# ---------------------------------------------------------------------------
def get_columns():
	return [
		{"label": _("Project"),              "fieldname": "project",             "fieldtype": "Link",     "options": "Project",         "width": 160},
		{"fieldname": "project_name",        "label": _("Project Name"),         "fieldtype": "Data",                                   "width": 220},
		{"label": _("Department"),           "fieldname": "department",          "fieldtype": "Link",     "options": "Department",      "width": 140},
		{"label": _("Branch"),               "fieldname": "branch",              "fieldtype": "Link",     "options": "Branch",          "width": 130},
		{"label": _("Employment Type"),      "fieldname": "employment_type",     "fieldtype": "Link",     "options": "Employment Type", "width": 140},
		{"label": _("Month"),                "fieldname": "month",               "fieldtype": "Data",                                   "width": 110},
		{"label": _("Stand By"), "fieldname": "stand_by", "fieldtype": "Data", "width": 100},
		{"label": _("Working Hours"),        "fieldname": "working_hours",       "fieldtype": "Float",    "precision": 2,               "width": 130},
		{"label": _("OT Hours"),             "fieldname": "ot_hours",            "fieldtype": "Float",    "precision": 2,               "width": 110},
		{"label": _("Working Hour Amount"),  "fieldname": "working_hour_amount", "fieldtype": "Currency",                               "width": 170},
		{"label": _("OT Hour Amount"),       "fieldname": "ot_hour_amount",      "fieldtype": "Currency",                               "width": 150},
	]


# ---------------------------------------------------------------------------
# Data
# ---------------------------------------------------------------------------
def get_data(filters):
	conditions, values = get_conditions(filters)

	query = """
		SELECT
			IFNULL(tsd.project, '(No Project)')                              AS project,
			proj.project_name                                                AS project_name,
			IFNULL(IFNULL(ts.department, emp.department), '')                AS department,
			IFNULL(emp.branch, '')                                           AS branch,
			IFNULL(emp.employment_type, '')                                  AS employment_type,
			DATE_FORMAT(ts.start_date, '%%b-%%Y')                            AS month,
			MONTH(ts.start_date)                                             AS month_num,
			YEAR(ts.start_date)                                              AS year_num,
			CASE WHEN MAX(ts.stand_by) = 1 THEN 'Yes' ELSE 'No' END          AS stand_by,

			SUM(IFNULL(tsd.working_hours, 0))                                AS working_hours,
			SUM(IFNULL(tsd.ot_hrs, 0))                                       AS ot_hours,

			SUM(IFNULL(tsd.working_hours, 0) * IFNULL(tsd.working_hour_rate_, 0))
			                                                                 AS working_hour_amount,
			SUM(IFNULL(tsd.ot_hrs, 0) * IFNULL(tsd.ot_hours_rate_, 0))
			                                                                 AS ot_hour_amount
		FROM
			`tabTimesheet Detail` tsd
		INNER JOIN `tabTimesheet` ts
			ON ts.name = tsd.parent
		LEFT JOIN `tabEmployee` emp
			ON emp.name = ts.employee
		LEFT JOIN
			`tabProject` proj ON proj.name = tsd.project
		WHERE
			ts.docstatus != 2
			{conditions}
		GROUP BY
			tsd.project,
			proj.project_name,
			IFNULL(ts.department, emp.department),
			emp.branch,
			emp.employment_type,
			YEAR(ts.start_date),
			MONTH(ts.start_date)
		ORDER BY
			tsd.project ASC,
			IFNULL(ts.department, emp.department) ASC,
			YEAR(ts.start_date) ASC,
			MONTH(ts.start_date) ASC
	""".format(conditions=conditions)

	data = frappe.db.sql(query, values, as_dict=True)
	# return add_group_totals(data)
	return data

# ---------------------------------------------------------------------------
# Filters → SQL conditions
# ---------------------------------------------------------------------------
def get_conditions(filters):
	conditions = []
	values = {}

	if filters.get("month"):
		month_num = [
			"January", "February", "March", "April", "May", "June",
			"July", "August", "September", "October", "November", "December"
		].index(filters["month"]) + 1
		conditions.append("AND MONTH(ts.start_date) = %(month_num)s")
		values["month_num"] = month_num

	if filters.get("year"):
		conditions.append("AND YEAR(ts.start_date) = %(year)s")
		values["year"] = filters["year"]

	if filters.get("project"):
		conditions.append("AND tsd.project = %(project)s")
		values["project"] = filters["project"]

	if filters.get("department"):
		conditions.append("AND IFNULL(ts.department, emp.department) = %(department)s")
		values["department"] = filters["department"]

	if filters.get("branch"):
		conditions.append("AND emp.branch = %(branch)s")
		values["branch"] = filters["branch"]

	if filters.get("employment_type"):
		conditions.append("AND emp.employment_type = %(employment_type)s")
		values["employment_type"] = filters["employment_type"]

	if filters.get("state"):
		conditions.append("AND ts.status = %(state)s")
		values["state"] = filters["state"]

	if filters.get("stand_by"):
		if filters["stand_by"] == "Yes":
			conditions.append("AND ts.stand_by = 1")   
		elif filters["stand_by"] == "No":
			conditions.append("AND ts.stand_by = 0") 

	return " ".join(conditions), values


# ---------------------------------------------------------------------------
# Group subtotals
# ---------------------------------------------------------------------------
# def add_group_totals(data):
# 	"""
# 	- Shows project name only on the FIRST row of each group (blank for rest).
# 	- Appends a bold subtotal row at the end of each group (when > 1 row).
# 	"""
# 	if not data:
# 		return data

# 	result       = []
# 	current_proj = None
# 	group_rows   = []

# 	def flush_group(rows):
# 		# Blank out the project on every row except the first
# 		for i, row in enumerate(rows):
# 			if i > 0:
# 				row["project"] = ""

# 		result.extend(rows)

# 		# Only add a Total row when the group has more than one detail row
# 		if len(rows) > 1:
# 			result.append({
# 				"project":             rows[0].get("_project_key"),   # bold label
# 				"department":          "",
# 				"branch":              "",
# 				"employment_type":     "",
# 				"month":               _("Total"),
# 				"working_hours":       sum(r.get("working_hours")       or 0 for r in rows),
# 				"ot_hours":            sum(r.get("ot_hours")            or 0 for r in rows),
# 				"working_hour_amount": sum(r.get("working_hour_amount") or 0 for r in rows),
# 				"ot_hour_amount":      sum(r.get("ot_hour_amount")      or 0 for r in rows),
# 				"bold": 1,
# 			})

# 	for row in data:
# 		proj_key = row["project"]  # original value before we blank it
# 		row["_project_key"] = proj_key  # stash for the total row label

# 		if proj_key != current_proj:
# 			if group_rows:
# 				flush_group(group_rows)
# 			current_proj = proj_key
# 			group_rows   = []

# 		group_rows.append(row)

# 	if group_rows:
# 		flush_group(group_rows)

	# return result



# ---------------------------------------------------------------------------
# Approval Method - Mark timesheets as approved
# ---------------------------------------------------------------------------
@frappe.whitelist()
def approve_timesheets(filters):
	"""
	Approve all Timesheets matching the given filters.
	Only users with 'General manager' role can execute this.
	"""
	# Check if user has General manager role (note: lowercase 'm')
	if not frappe.has_permission("Timesheet", "write") or not "General manager" in frappe.get_roles():
		frappe.throw(_("You don't have permission to approve timesheets."))
	
	# Parse filters if they come as JSON string
	if isinstance(filters, str):
		import json
		filters = json.loads(filters)
	
	# Build conditions to find matching timesheets
	conditions = []
	values = {}
	
	if filters.get("project"):
		conditions.append("tsd.project = %(project)s")
		values["project"] = filters["project"]
	
	if filters.get("department"):
		conditions.append("IFNULL(ts.department, emp.department) = %(department)s")
		values["department"] = filters["department"]
	
	if filters.get("branch"):
		conditions.append("emp.branch = %(branch)s")
		values["branch"] = filters["branch"]
	
	if filters.get("employment_type"):
		conditions.append("emp.employment_type = %(employment_type)s")
		values["employment_type"] = filters["employment_type"]
	
	if filters.get("month"):
		month_num = [
			"January", "February", "March", "April", "May", "June",
			"July", "August", "September", "October", "November", "December",
		].index(filters["month"]) + 1
		conditions.append("MONTH(ts.start_date) = %(month_num)s")
		values["month_num"] = month_num
	
	if filters.get("year"):
		conditions.append("YEAR(ts.start_date) = %(year)s")
		values["year"] = filters["year"]
	
	if filters.get("state"):
		conditions.append("ts.status = %(state)s")
		values["state"] = filters["state"]
	
	if filters.get("stand_by"):
		if filters["stand_by"] == "Yes":
			conditions.append("ts.stand_by = 1")
		elif filters["stand_by"] == "No":
			conditions.append("ts.stand_by = 0")
	
	where_clause = " AND ".join(conditions) if conditions else "1=1"
	
	# Get all matching timesheets (distinct parent timesheets)
	timesheets = frappe.db.sql(
		f"""
		SELECT DISTINCT ts.name
		FROM `tabTimesheet Detail` tsd
		INNER JOIN `tabTimesheet` ts ON ts.name = tsd.parent
		LEFT JOIN `tabEmployee` emp ON emp.name = ts.employee
		WHERE ts.docstatus != 2
		AND {where_clause}
		""",
		values,
		as_dict=True
	)
	
	if not timesheets:
		frappe.msgprint(_("No timesheets found matching the selected filters."))
		return {"count": 0}
	
	# Update each timesheet
	count = 0
	for ts in timesheets:
		try:
			doc = frappe.get_doc("Timesheet", ts.name)
			doc.custom_approved = 1
			doc.save(ignore_permissions=True)
			count += 1
		except Exception as e:
			frappe.log_error(f"Error approving timesheet {ts.name}: {str(e)}", "Timesheet Approval Error")
	
	frappe.db.commit()
	
	return {"count": count}

File: mkan_customization/mkan_customization/report/timesheet_monthly_summary_per_project/timesheet_monthly_summary_per_project.js
frappe.query_reports["Timesheet Monthly Summary Per Project"] = {
	filters: [
		{
			fieldname: "month",
			label: __("Month"),
			fieldtype: "Select",
			options: [
				"",
				"January","February","March","April","May","June",
				"July","August","September","October","November","December",
			].join("\n"),
			width: "100",
		},
		{
			fieldname: "year",
			label: __("Year"),
			fieldtype: "Int",
			default: frappe.datetime.now_datetime().substring(0, 4),
			width: "80",
		},
		{
			fieldname: "project",
			label: __("Project"),
			fieldtype: "Link",
			options: "Project",
			width: "100",
		},
		{
			fieldname: "department",
			label: __("Department"),
			fieldtype: "Link",
			options: "Department",
			width: "100",
			get_query: function () {
				return { filters: { company: frappe.defaults.get_default("company") } };
			},
		},
		{
			fieldname: "branch",
			label: __("Branch"),
			fieldtype: "Link",
			options: "Branch",
			width: "100",
		},
		{
			fieldname: "employment_type",
			label: __("Employment Type"),
			fieldtype: "Link",
			options: "Employment Type",
			width: "100",
		},
		{
			fieldname: "state",
			label: __("State"),
			fieldtype: "Select",
			options: ["", "Draft", "Submitted", "Cancelled", "Payslip", "Completed"].join("\n"),
			width: "100",
		},
		{
			fieldname: "stand_by",
			label: __("Stand By"),
			fieldtype: "Select",
			options: "\nYes\nNo",
		},
	],

	onload: function(report) {
		console.log("Report onload called");
		console.log("User roles:", frappe.user_roles);
		console.log("Has General Manager role:", frappe.user.has_role("General manager"));
		
		// Check if user has General manager role (note: lowercase 'm')
		if (frappe.user.has_role("General manager")) {
			console.log("Adding Approval button");
			report.page.add_inner_button(__("Approval"), function() {
				let filters = report.get_values();
				
				// Validate that required filters are set
				if (!filters.year || !filters.month) {
					frappe.msgprint({
						title: __("Missing Filters"),
						message: __("Please select Project, Year, and Month before approving."),
						indicator: "red"
					});
					return;
				}

				// Confirm approval action
				frappe.confirm(
					__("Are you sure you want to approve all Timesheets for the selected filters?"),
					function() {
						// Call server method to approve timesheets
						frappe.call({
							method: "mkan_customization.mkan_customization.report.timesheet_monthly_summary_per_project.timesheet_monthly_summary_per_project.approve_timesheets",
							args: {
								filters: filters
							},
							freeze: true,
							freeze_message: __("Approving timesheets..."),
							callback: function(r) {
								if (r.message) {
									frappe.msgprint({
										title: __("Success"),
										message: __("{0} timesheet(s) have been approved successfully.", [r.message.count]),
										indicator: "green"
									});
									// Refresh the report
									report.refresh();
								}
							},
							error: function(r) {
								frappe.msgprint({
									title: __("Error"),
									message: __("Failed to approve timesheets. Please try again."),
									indicator: "red"
								});
							}
						});
					}
				);
			});
		} else {
			console.log("User does not have General manager role");
		}
	},

	// -------------------------------------------------------------------------
	// Formatter
	// -------------------------------------------------------------------------
	formatter: function (value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);

		if (data && data.bold) {
			value = `<span style="font-weight:700;">${value}</span>`;
		}
		if (data && !data.bold) {
			if (column.fieldname === "working_hours" && data.working_hours === 0) {
				value = `<span style="color:#aaa;">${value}</span>`;
			}
			if (column.fieldname === "ot_hours" && data.ot_hours > 0) {
				value = `<span style="">${value}</span>`;
			}
		}
		return value;
	},

	// -------------------------------------------------------------------------
	// Chart — 4 datasets: Working Hours, OT Hours, Working Amount, OT Amount
	// Uses two Y-axes: hours (left) and amounts (right via scale label trick)
	// -------------------------------------------------------------------------
	get_chart_data: function (columns, result) {
		// Skip Total/bold rows and collect unique months in order
		const months    = [];
		const monthSeen = {};

		result.forEach(function (row) {
			if (row.month && row.month !== __("Total") && !row.bold && !monthSeen[row.month]) {
				months.push(row.month);
				monthSeen[row.month] = true;
			}
		});

		function sumByMonth(field) {
			return months.map(function (m) {
				return result
					.filter(r => r.month === m && !r.bold)
					.reduce((acc, r) => acc + (r[field] || 0), 0);
			});
		}

		return {
			data: {
				labels: months,
				datasets: [
					{
						name:      __("Working Hours"),
						values:    sumByMonth("working_hours"),
						chartType: "bar",
					},
					{
						name:      __("OT Hours"),
						values:    sumByMonth("ot_hours"),
						chartType: "bar",
					},
					{
						name:      __("Working Hour Amount"),
						values:    sumByMonth("working_hour_amount"),
						chartType: "bar",
					},
					{
						name:      __("OT Hour Amount"),
						values:    sumByMonth("ot_hour_amount"),
						chartType: "bar",
					},
				],
			},
			type: "axis-mixed",
			barOptions:  { stacked: false },
			lineOptions: { regionFill: 0, hideDots: 0 },
			colors: ["#fa5d77ff", "#4da6ff", "#00c49f", "#ff7f50"],
		};
	},
};